### ISSUE TITLE
[Bug][Attention] temperature=0 identical requests diverge when `sliding_window` is set in config.json

### ISSUE THREAD

**Issue by whthust** [vllm-project/vllm-ascend#16704]
## Summary

- **Problem**: with `temperature=0`, identical prompts sent concurrently (8 or 16 requests per round, multiple rounds) produce **different outputs** — every divergence is a tie-point argmax flip where the top-2 candidate logprobs are exactly equal (≤1 bf16 ulp). No semantic/computation error is involved.
- **Root cause**: when the checkpoint's `config.json` carries a `sliding_window` field, the model's decode path splits into **two different FIA call forms depending on the batch state of the step**:
  - **Decode-only steps** (`DecodeOnly` state): all 62 layers route to `_forward_fia_slidingwindow` — BSH layout `(N, 1, 1536)`, **no causal mask**, `pre_tokens=sliding_window`.
  - **Mixed prefill+decode steps** (`ChunkedPrefill` state): all layers route to the generic TND `npu_fused_infer_attention_score` — packed layout, **causal 2048 triu mask**, `sparse_mode=3`.
  - The two call forms compute mathematically correct attention (the window never truncates), but **are not bitwise aligned** (~1 bf16 ulp per call; measured `0.000244 = 2⁻¹²` on rank-1 layer-23 attention output).
- **Effect**: the same logical token is computed by different kernels depending on whether its decode lands in a mixed step or a decode-only step. The ulp-level difference is amplified through the residual stream (layer 23: `2⁻¹²` → layer 61: `4.0`) and flips argmax at exact logprob ties → concurrent identical requests diverge.
- **Verified fix**: removing `sliding_window` from `config.json` (or `--disable-sliding-window`) unifies every step to the TND path → **bitwise-identical outputs at concurrency 8/16, and latency drops from 73 s → 37 s** (the sliding-window BSH path was itself the slow path).

## Root cause (code chain)

```notranslate
config.json: "sliding_window": 131072          # == max_position_embeddings, i.e. full attention
  → hf_text_config.sliding_window              (vllm/config/model.py:1115-1117, get_sliding_window)
  → cache_config.sliding_window                (vllm/engine/arg_utils.py:1524)
  → Attention.sliding_window ≠ None            (vllm/model_executor/layers/attention/attention.py:212-217,
                                                falls back to cache_config when per_layer is None)
  → AscendAttentionBackendImpl.sliding_window ≠ None

vllm_ascend/attention/attention_v1.py:775-781 (branch):
    if (attn_state == DecodeOnly
        and self.sliding_window is not None
        and seq_lens.shape[0] == query.size(0)
        and self.sinks is None):
        return self._forward_fia_slidingwindow(...)     # taken by EVERY pure decode step
    # otherwise → generic TND FIA at line 821

_forward_fia_slidingwindow (attention_v1.py:730-758):
    npu_fused_infer_attention_score(
        query.view(B, 1, 1536),            # input_layout="BSH"
        pre_tokens=131072,                 # = sliding_window
        # NO atten_mask, no sparse_mode
    )

generic path (attention_v1.py:821-835):
    npu_fused_infer_attention_score(
        query,                             # input_layout="TND"
        atten_mask=<2048×2048 causal triu>, # attention_mask.py:54-59 (shared cached mask)
        sparse_mode=3,
    )
```

The same operator (`npu_fused_infer_attention_score`) is invoked with two different layouts/masks/sparse modes, so internally **different kernel/tiling implementations** run for the same logical computation, producing different reduction orders → ~1 bf16 ulp of difference per call.

**Failure flow:**

```notranslate
request trajectory (concurrent traffic, P-D mixing is unavoidable)
   │
   ├─ decode-only step (BSH sliding-window kernel)   ─┐
   ├─ decode-only step (BSH)                          │  same token,
   ├─ mixed step (TND generic kernel)  <── switch!    │  different kernel
   ├─ decode-only step (BSH)                          │  → ~1 ulp gap
   ▼                                                 ─┘
residual stream amplifies: layer23 0.000244 → o_proj 0.000977
   → all-reduce broadcast 0.003906 → layer61 4.0
   ▼
logits top-2 exactly tied (e.g. '为'/'为主要' both = -1.028142)
   ▼
greedy argmax flips → temperature=0 outputs diverge
```

**Why pure-decode is "clean" and mixing causes the problem**: decode-only trajectories use the BSH form uniformly (self-consistent); mixed steps force a fallback to the TND form (the sliding-window branch requires `DecodeOnly` state and exactly 1 query token per request). Each path is individually correct — the **inconsistency between the two paths for the same token** is the bug trigger.

## Environment

- **vllm-ascend version**: 0.18.0rc1 (fork)
- **Model**: custom 62-layer dense Llama (hidden 6144, bf16, TP=4), multimodal wrapper
- **Hardware**: Ascend 910B × 4

**Serving configuration (production, anonymized)** — this is the exact deployment that hit the issue; the reproduction used the same configuration **without** the `--disable-sliding-window` flag (that flag is the workaround we ended with):

```bash
# environment variables
export VLLM_USE_MODELSCOPE=True
export VLLM_WORKER_MULTIPROC_METHOD=spawn
export PYTORCH_NPU_ALLOC_CONF=max_split_size_mb:256
export VLLM_USE_V1=1
export VLLM_LOGGING_LEVEL=INFO
export HCCL_DETERMINISTIC=true
export LCCL_DETERMINISTIC=1
export CLOSE_MATMUL_K_SHIFT=1
export VLLM_BATCH_INVARIANT=1
export VLLM_ASCEND_ENABLE_NZ=0
export ATB_MATMUL_SHUFFLE_K_ENABLE=0
export ATB_LLM_LCOC_ENABLE=0
export PYTHONHASHSEED=42
export PYTORCH_DETERMINISTIC=1
export ASCEND_LAUNCH_BLOCKING=1
export TORCH_NPU_ENABLE_DETERMINISTIC=1
export ASCEND_VISIBLE_DEVICES=0,1,2,3

# engine launch (paths/model name anonymized)
vllm serve <MODEL_PATH> \
    --host 0.0.0.0 --port 9998 \
    --served-model-name <custom-model-name> \
    --tensor_parallel_size 4 \
    --gpu-memory-utilization 0.85 \
    --max-num-seqs 32 \
    --max-model-len 131072 \
    --max-num-batched-tokens 131072 \
    --block-size 128 \
    --trust-remote-code \
    --enforce-eager \
    --no-enable-chunked-prefill \
    --no-enable-prefix-caching \
    --no-async-scheduling \
    --enable-auto-tool-choice \
    --tool-call-parser <tool-parser> \
    --chat-template <chat-template-path> \
    --allowed-local-media-path <local-media-path> \
    --limit-mm-per-prompt '{"video": 1}' \
    --media-io-kwargs '{"video": {"num_frames": -1, "fps": 1, "max_frames": 32, "min_frames": 2}}' \
    --disable-sliding-window   # <- the verified workaround (absent when the bug reproduced)
```

Note: chunked prefill is disabled in our deployment, yet **P-D mixed steps still occur** (multiple requests scheduled together make prefill and decode tokens share one step), which is exactly the trigger condition.

## Reproduction

1. **Online**: `temperature=0, top_p=1, top_k=0`, send the same long-text prompt with 8-process concurrency × 3 rounds → **2~7 distinct results per round**; every first-divergence token shows an exact top-2 logprob tie (e.g., both candidates `-1.028142`), i.e. a tie-point flip.
2. **Offline in-situ** (hook capture, 5 reproductions): one probe token decoded once in a giant mixed step (1010-token prefill + 2 decode tokens in one batch) and once in a pure decode step:
   - `rank1 layer-23 attention output diff = 0.000244` with q/k/v, KV contents, mask, and all upstream layers (0-22) **bitwise identical**; `rank0/2/3` identical.
   - The diff propagates: `o_proj 0.000977 → post-all-reduce 0.003906 → layer 61 diff = 4.0`.
3. **Call fingerprint** (all 62 FIA calls per step): decode-only steps → all `(1, 1, 1536)` BSH form, real `kv_len`, no mask; mixed step → all `(1012, 12, 128)` TND form, causal mask.
4. **End-to-end offline repro (decisive)**: replay the operator with the engine's real KV/Query, changing **only the call form** (BSH sliding-window vs TND):
   - `rank1: equal=False max_diff=0.000244`; `rank0/2/3: equal=True`
   - Identical value and identical rank pattern as the in-engine measurement — mechanism proven offline.

## Exclusion evidence (all measured, none inferred)

- matmul/`F.linear` at all relevant shapes (incl. TP=4 per-rank `K=1536`): batch-invariant ✓
- TP all-reduce across message shapes (`M=1/3/34` vs `1012`): invariant ✓
- rope (per-row, no cross-row reduction): invariant ✓
- custom-op wrapper `vllm::unquantized_gemm`: registration clean; single-process `custom == F.linear == explicit batch-invariant op` ✓
- prefill computation: mixed (33 rows) vs pure (32 rows), **62 layers × 32 rows × 4 ranks bitwise identical** ✓
- attention mask: shared cached `triu(2048)` for every step (code-level identical) ✓
- KV physical block id (V1) / block-table padding content (V2): insensitive ✓
- FIA batch-shape invariance (TND mixed vs TND solo): bitwise equal ✓
- Call-history / workspace state dependence (V3 + control): no effect ✓
- `batch_invariant_ops` wrapper is a pure function with no workspace handle ✓
- **Natural A/B control**: `qwen3-vl-30b` (no `sliding_window` in config) is fully consistent on the same stack — and its 3 exact tie points never flip (perturbation = 0).

## Proposed fix

1. **Config level (immediate, verified)**: remove `"sliding_window": 131072` from the checkpoint `config.json` (the value equals `max_position_embeddings`, i.e. a semantic no-op for any sequence < 131072), or start the server with **`--disable-sliding-window`** (`vllm/engine/arg_utils.py:448` → `vllm/config/model.py:631-634` sets `hf_text_config.sliding_window = None`).
   - ⚠️ `--hf-overrides '{"sliding_window": null}'` does **NOT** work here: the field is read from `hf_text_config` (`model.py:1115-1117`), which top-level HF overrides do not reach.
2. **Framework level**: skip the sliding-window branch when `sliding_window >= max_position_embeddings` (the window never truncates, i.e. full attention — taking the BSH path buys nothing but inconsistency and latency), or make the sliding-window branch numerically identical to the TND path; and document that setting `sliding_window` switches the decode kernel path (affecting cross-trajectory numerical consistency).
3. **Operator level (CANN)**: make the BSH sliding-window form and the TND form of `npu_fused_infer_attention_score` **row-wise bitwise consistent** for identical inputs (including their batch-invariant variants). Reproduction assets available on request (captured FIA args + replay script).

## Validation

- **After the config fix**: concurrency 8 × 2 rounds and 16 × 3 rounds — **all outputs bitwise identical**; latency `73 s → 37 s` per round (the sliding-window BSH path was itself the slow path).
- **Before**: 2~7 distinct outputs per round at concurrency 8; **after**: 1 identical output at concurrency 8 and 16.
- No semantic change for sequences ≤ 131072 (the window never truncates); KV-cache spec becomes `FullAttentionSpec`, identical under `max_model_len 8192`.
- A scheduler-side workaround (forbid P-D mixing) also achieves consistency at +12~15 % latency, confirming the P-D mixed step as the trigger point.

## Compatibility notes

- **Do NOT remove the field for genuinely SWA-trained models** (sliding-window semantics for long contexts). For those, fix at framework/operator level instead (path unification or numeric alignment).
- If a checkpoint family shares the config template (`sliding_window = max_position`), all siblings likely carry the same landmine and need the same treatment.
- This issue was investigated and written with AI assistance; all experiments, measurements, and the final fix were executed and verified on our side by a human engineer.

## Attachments (available on request)

- `fia_replay.log`: end-to-end V4 reproduction (BSH vs TND, rank1 `0.000244`).
- `fia_calls_rank{0..3}.pt`: full FIA call fingerprints per step (BSH decode-only vs TND mixed).
- `fia_dump_rank{0..3}.pt`: captured engine FIA args incl. ground-truth outputs (replay faithfulness == engine).
- `issue-analysis.md`: complete 29-stage investigation archive with all raw logs.
